[native] use velox's cmake `ResolveDependency` macros

nkaretnikov opened the pull request:
Fixes #18685.

nkaretnikov commented on presto-native-execution/CMakeLists.txt:
```diff
@@ -14,7 +14,8 @@ cmake_minimum_required(VERSION 3.10)
 
-set(VELOX_ROOT ${CMAKE_BINARY_DIR}/velox)
+# NB: uses CMAKE_SOURCE_DIR to support out-of-source builds
+set(VELOX_ROOT ${CMAKE_SOURCE_DIR}/velox)
```
Unrelated change: wanted to support out-of-source builds when testing. Will remove before submitting.

nkaretnikov commented on presto-native-execution/CMakeLists.txt:
```diff
@@ -23,6 +24,21 @@ execute_process(
+    CACHE
+      STRING
+      "Default source for all dependencies with source builds enabled: AUTO SYSTEM BUNDLED."
+)
```
These are used internally in ResolveDependency macros (copied from there).

nkaretnikov commented on presto-native-execution/CMakeLists.txt:
```diff
@@ -58,8 +74,6 @@ option(PRESTO_ENABLE_REMOTE_FUNCTIONS "Enable remote function support" OFF)
 option(PRESTO_ENABLE_TESTING "Enable tests" ON)
 
 # Set all Velox options below
-add_compile_definitions(FOLLY_HAVE_INT128_T=1)
```
Moved below - same as in velox's CMakeLists.txt

nkaretnikov commented on presto-native-execution/CMakeLists.txt:
```diff
@@ -103,81 +117,88 @@ set(VELOX_BUILD_TEST_UTILS
+    thread)
+
+set_source(Boost)
+resolve_dependency(Boost 1.66.0 COMPONENTS ${BOOST_INCLUDE_LIBRARIES})
```
Same as in velox.

v-jizhang replied:
`headers ` is not needed.

nkaretnikov commented on presto-native-execution/CMakeLists.txt:
```diff
@@ -103,81 +117,88 @@ set(VELOX_BUILD_TEST_UTILS
-if(BZIP2_FOUND)
-  list(APPEND FOLLY_WITH_DEPENDENCIES ${BZIP2_LIBRARIES})
-endif()
-include_directories(SYSTEM ${FOLLY_INCLUDE_DIRS})
```
Same as in velox, see https://github.com/facebookincubator/velox/pull/4735 (also removes BZIP2 check there)

nkaretnikov commented on presto-native-execution/CMakeLists.txt:
```diff
@@ -103,81 +117,88 @@ set(VELOX_BUILD_TEST_UTILS
+# to use NOT TARGET in the last if block and move two `set` calls above it
+# set_source(re2)
+# resolve_dependency(re2)
 find_library(RE2 re2)
```
Will submit a fix to velox once this PR is approved.

v-jizhang replied:
Should we add new files like `double-conversion.cmake` etc. to Presto only? They won't be used by velox.

nkaretnikov commented on presto-native-execution/presto_cpp/main/CMakeLists.txt:
```diff
@@ -64,10 +64,10 @@ target_link_libraries(
-  ${GLOG}
-  ${GFLAGS_LIBRARIES}
+  glog::glog
+  gflags::gflags
```
Here and elsewhere: same as in velox.

nkaretnikov commented:
Summary:
- this is WIP - to get early feedback
- left some notes to help navigate the code
- I've only made sure this builds in docker and passes the tests (not reviewed, haven't checked if libs build statically or dynamically, etc.)
- 2 tests fail for me locally, but these also fail with an upstream cmake build - I haven't investigated - attached below
- NB: velox itself doesn't use `resolve_dependency` for everything

Tests that fail locally when you build via cmake:
```
          1 - presto_operators_test (SEGFAULT)
          7 - presto_server_test (Failed)
```

Dockerfile with my branch to make sure it builds in a clean env.
For dev: I pass presto repo from host via `-v`.

```Dockerfile
# docker build --tag presto:latest - < Dockerfile && docker run -it presto:latest /bin/bash

FROM ubuntu:22.04

ARG DEBIAN_FRONTEND=noninteractive

RUN apt-get update
RUN apt-get install -y git build-essential cmake sudo python3-pip unzip

# Uses my fork for now
RUN git clone -b resolve-dependency-cmake https://github.com/nkaretnikov/presto && \
  cd presto/presto-native-execution && \
  make submodules && \
  bash ./scripts/setup-ubuntu.sh && \
  mkdir _build && \
  cd _build && \
  cmake .. && \
  cmake --build . --config Release -j$(($(nproc) * 3/4)) && \
  ctest || true  # 2 tests fail (same as with the upstream cmake build)
```

nkaretnikov commented:
@v-jizhang @kgpai As discussed, please take a look at this early draft.

v-jizhang commented on presto-native-execution/CMakeLists.txt:
```diff
@@ -23,6 +24,21 @@ execute_process(
+# See velox/CMake/resolve_dependency_modules/README.md
+list(PREPEND CMAKE_MODULE_PATH "${VELOX_ROOT}/CMake"
+     "${VELOX_ROOT}/CMake/third-party")
+include(${VELOX_ROOT}/CMake/ResolveDependency.cmake)
```
`include(ResolveDependency)` should work here because `CMAKE_MODULE_PATH `is set by previous line.

v-jizhang commented on presto-native-execution/CMakeLists.txt:
```diff
@@ -103,81 +117,88 @@ set(VELOX_BUILD_TEST_UTILS
+else()
+  set(glog_SOURCE SYSTEM)
+endif()
+resolve_dependency(glog)
```
Any reason why there is no` set_source(glog)`?

nkaretnikov commented on presto-native-execution/CMakeLists.txt:
```diff
@@ -155,12 +155,6 @@ resolve_dependency(glog)
-find_library(EVENT event)
-
-# TODO
-find_library(DOUBLE_CONVERSION double-conversion)
```
event and double-conversion appear to be not needed at all - I can build locally without them

nkaretnikov commented on presto-native-execution/CMake/Findantlr4-runtime.cmake:
```diff
@@ -0,0 +1,21 @@
+    add_library(antlr4-runtime::antlr4-runtime UNKNOWN IMPORTED)
+    set_target_properties(antlr4-runtime::antlr4-runtime PROPERTIES INTERFACE_INCLUDE_DIRECTORIES "${ANTLR4_RUNTIME_INCLUDE_DIRS}")
+    set_target_properties(antlr4-runtime::antlr4-runtime PROPERTIES IMPORTED_LINK_INTERFACE_LANGUAGES "C" IMPORTED_LOCATION "${ANTLR4_RUNTIME_LIBRARIES}")
+  endif()
```
The first half of this file is what we had in CMakeLists.txt. The rest is similar to Find modules in velox.

We define our own module because it's not provided/used by velox.

nkaretnikov commented on presto-native-execution/CMake/Findre2.cmake:
```diff
@@ -0,0 +1,29 @@
+set(re2_LIBRARIES ${re2_lib})
+set(re2_INCLUDE_DIRS ${re2_include})
+
+if (NOT TARGET re::re)
```
Typo. Needs to be re2. Will fix.

nkaretnikov commented on presto-native-execution/CMake/Findre2.cmake:
```diff
@@ -0,0 +1,29 @@
+  target_link_libraries(re2::re2 INTERFACE ${re2_LIBRARIES})
+  set_target_properties(re2::re2 PROPERTIES IMPORTED_LOCATION
+                                            "${re2_LIBRARIES}")
+endif()
```
This is a copy of the re2 file from velox, with the if/set logic changed, which doesn't error out if re2 is processed twice.

nkaretnikov commented on presto-native-execution/CMakeLists.txt:
```diff
@@ -119,6 +121,19 @@ set(VELOX_BUILD_TEST_UTILS
+# allows us to bundle the same Find modules that are shipped with velox, but
+# with our own modifications. For instance, our modules can detect if the target
+# dependency is already defined.
+add_subdirectory(velox)
```
This allows us to process all of velox's Find modules first before loading ours. Because both velox and presto depend on re2, we process the Find module for re2 twice. However, the upstream one is written in a way that complains if you use it twice (unlike other Find modules in the velox repo). I provide a copy of re2 with a patch in this repo that fixes it.

nkaretnikov commented on presto-native-execution/CMakeLists.txt:
```diff
@@ -135,13 +150,8 @@ resolve_dependency(Boost 1.66.0 COMPONENTS ${BOOST_INCLUDE_LIBRARIES})
-  PATH_SUFFIXES antlr4-runtime)
-include_directories(SYSTEM ${ANTLR4_RUNTIME_INCLUDE_DIR})
+set_source(antlr4-runtime)
+resolve_dependency(antlr4-runtime)
```
Can be imported via find_module now because I bundle it with presto.

nkaretnikov commented on presto-native-execution/CMakeLists.txt:
```diff
@@ -177,8 +187,9 @@ set(OPENSSL_ROOT_DIR "${OPT_OPENSSL_DIR}/opt/openssl@1.1")
-find_library(LIBSODIUM_LIBRARY NAMES sodium)
+# Note: the uppercase name here is correct.
+set_source(Sodium)
+resolve_dependency(Sodium)
```
This didn't work before because the name defined in the Find module uses uppercase (bundled with velox).

nkaretnikov commented on presto-native-execution/CMakeLists.txt:
```diff
@@ -188,10 +199,8 @@ resolve_dependency(fizz)
-# set_source(re2) resolve_dependency(re2)
-find_library(RE2 re2)
+set_source(re2)
+resolve_dependency(re2)
```
Solved by bundling a patched re2 with presto.

nkaretnikov commented on presto-native-execution/CMakeLists.txt:
```diff
@@ -202,17 +211,7 @@ set(PROXYGEN_LIBRARIES ${PROXYGEN_HTTP_SERVER} proxygen::proxygen
-# set this for backwards compatibility, will be overwritten in velox/
-set(VELOX_GTEST_INCUDE_DIR "velox/third_party/googletest/googletest/include")
-
-add_subdirectory(velox)
```
Moved to the start of the file, see the comment there.

nkaretnikov commented on presto-native-execution/CMakeLists.txt:
```diff
@@ -32,15 +32,6 @@ list(PREPEND CMAKE_MODULE_PATH "${CMAKE_CURRENT_SOURCE_DIR}/CMake"
-      STRING
-      "Default source for all dependencies with source builds enabled: AUTO SYSTEM BUNDLED."
-)
-
```
No longer needed after I've moved things around.

nkaretnikov commented on presto-native-execution/presto_cpp/main/http/CMakeLists.txt:
```diff
@@ -20,14 +20,16 @@ target_link_libraries(
   velox_exception
   ${PROXYGEN_LIBRARIES}
+  fizz::fizz
+  wangle::wangle
```
fizz and wangle were part of PROXYGEN_LIBRARIES before, but I didn't want to have them defined in the Find module, so they are now here.

nkaretnikov commented on presto-native-execution/CMakeLists.txt:
```diff
@@ -103,95 +110,76 @@ set(VELOX_BUILD_TEST_UTILS
-find_library(LZO lzo2)
-find_library(ZSTD zstd)
-find_package(ZLIB)
-find_library(SNAPPY snappy)
```
These appear to be velox's dependencies, so dropped these.

nkaretnikov commented:
@v-jizhang @kgpai Cleaned up, code now uses `resolve_dependency`. Please take a look.

majetideepak commented:
@nkaretnikov thanks for taking up this. Is BUNDLED mode supported in this PR?

kgpai commented on presto-native-execution/presto_cpp/main/http/filters/CMakeLists.txt:
```diff
@@ -13,7 +13,8 @@
                          StatsFilter.cpp)
 
-target_link_libraries(http_filters presto_common ${PROXYGEN_LIBRARIES})
+target_link_libraries(http_filters presto_common ${PROXYGEN_LIBRARIES}
```
Is there no proxygen::proxygen target ?

nkaretnikov replied:
I'll add an additional target for the proxygen http server to have more control over linking, then folks can use target names.

nkaretnikov commented on presto-native-execution/CMakeLists.txt:
```diff
@@ -103,95 +110,76 @@ set(VELOX_BUILD_TEST_UTILS
+# Set this for backwards compatibility, will be overwritten in velox/
+set(VELOX_GTEST_INCUDE_DIR "velox/third_party/googletest/googletest/include")
+include_directories(.)
+include_directories(${CMAKE_BINARY_DIR})
```
Fails to find `velox/velox/dwio/dwrf/proto/dwrf_proto.pb.h` otherwise.

In general, these includes were here before, I just moved them to the start of the file. I didn't do significant refactoring to keep this self-contained.

But now that you mention it, I'll see which can be removed.

nkaretnikov commented on presto-native-execution/CMakeLists.txt:
```diff
@@ -103,95 +110,76 @@ set(VELOX_BUILD_TEST_UTILS
+set_source(antlr4-runtime)
+resolve_dependency(antlr4-runtime)
+
+if(${gflags_SOURCE} STREQUAL "BUNDLED")
```
This is copied from velox. There's a comment explaining why it's this way, but real world implications of this are unknown to me. I'll remove this to keep things simple, should be easy to re-add if things break.

nkaretnikov commented on presto-native-execution/CMakeLists.txt:
```diff
@@ -103,95 +110,76 @@ set(VELOX_BUILD_TEST_UTILS
+set_source(Sodium)
+resolve_dependency(Sodium)
+set_source(fizz)
+resolve_dependency(fizz)
```
No.

`resolve_dependency` does work as expected, but `BUNDLED` uses `include`, which expects a `.cmake` file in `CMAKE_MODULE_PATH`.

velox doesn't provide this file for `fizz` and `wangle`. **Would you like me to add `.cmake` files for these two packages?**

Available `.cmake` files:

```
velox/CMake/resolve_dependency_modules/boost.cmake
velox/CMake/resolve_dependency_modules/fmt.cmake
velox/CMake/resolve_dependency_modules/folly.cmake
velox/CMake/resolve_dependency_modules/gflags.cmake
velox/CMake/resolve_dependency_modules/glog.cmake
velox/CMake/resolve_dependency_modules/gtest.cmake
velox/CMake/resolve_dependency_modules/icu.cmake
velox/CMake/resolve_dependency_modules/protobuf.cmake
velox/CMake/resolve_dependency_modules/pybind11.cmake
velox/CMake/resolve_dependency_modules/re2.cmake
velox/CMake/resolve_dependency_modules/simdjson.cmake
velox/CMake/resolve_dependency_modules/template.cmake
velox/CMake/resolve_dependency_modules/xsimd.cmake
```

nkaretnikov replied:
> Is BUNDLED mode supported in this PR?

@majetideepak see the thread above.

nkaretnikov commented on presto-native-execution/CMakeLists.txt:
```diff
@@ -103,95 +110,76 @@ set(VELOX_BUILD_TEST_UTILS
+set_source(fizz)
+resolve_dependency(fizz)
+set_source(wangle)
+resolve_dependency(wangle)
```
See the reply above regarding fizz.

nkaretnikov commented:
Updated:
- rebased on `master`
- added `proxygen::proxygenhttpserver` target, updated linking commands to use it
- removed `xxhash` from includes
- removed `if` from `glog`

nkaretnikov commented:
@mshang816 A gentle reminder to take a look at this one (sorry, I cannot assign you as a reviewer directly).

kgpai commented:
Looking.

kgpai reviewed (commented):
1) Running with proxygen_BUNDLED I get the following error:

```
-- Setting proxygen source to BUNDLED
CMake Error at velox/CMake/ResolveDependency.cmake:43 (include):
  include requested file is a directory:

    proxygen
Call Stack (most recent call first):
  velox/CMake/ResolveDependency.cmake:81 (build_dependency)
  CMakeLists.txt:168 (resolve_dependency)
```

2) Whats the right string to set antlr to bundled ? antlr4-runtime_SOURCE=BUNDLED isnt recognized by bash. 

3) Building with gflags_SOURCE=BUNDLED, gives me the following issues , (these might not be related to your cmake changes):
```
[622/679] Building CXX object presto_cpp/main/thrift/CMakeFiles/presto_thrift-cpp2-obj.dir/presto_cpp/main/thrift/gen-cpp2/presto_thrift_data.cpp.o
FAILED: presto_cpp/main/thrift/CMakeFiles/presto_thrift-cpp2-obj.dir/presto_cpp/main/thrift/gen-cpp2/presto_thrift_data.cpp.o 
ccache /Library/Developer/CommandLineTools/usr/bin/c++ -DFOLLY_HAVE_INT128_T=1 -I/Users/kpai/src/presto/presto-native-execution/. -I/Users/kpai/src/presto/presto-native-execution/_build/debug -I/Users/kpai/src/presto/presto-native-execution/velox -I/Users/kpai/src/presto/presto-native-execution/_build/debug/velox -I/Users/kpai/src/presto/presto-native-execution/_build/debug/_deps/gtest-src/googletest/include -isystem /usr/local/include -isystem /Users/kpai/homebrew/opt/openssl@1.1/include -isystem /usr/local -isystem /usr/local/include/antlr4-runtime -mcpu=apple-m1+crc -std=c++17 -fvisibility=hidden -fvisibility=hidden -fvisibility-inlines-hidden -Werror -Wno-nullability-completeness -Wno-deprecated-declarations -Wreorder -g -std=gnu++17 -arch arm64 -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX13.3.sdk -MD -MT presto_cpp/main/thrift/CMakeFiles/presto_thrift-cpp2-obj.dir/presto_cpp/main/thrift/gen-cpp2/presto_thrift_data.cpp.o -MF presto_cpp/main/thrift/CMakeFiles/presto_thrift-cpp2-obj.dir/presto_cpp/main/thrift/gen-cpp2/presto_thrift_data.cpp.o.d -o presto_cpp/main/thrift/CMakeFiles/presto_thrift-cpp2-obj.dir/presto_cpp/main/thrift/gen-cpp2/presto_thrift_data.cpp.o -c /Users/kpai/src/presto/presto-native-execution/_build/debug/presto_cpp/main/thrift/presto_cpp/main/thrift/gen-cpp2/presto_thrift_data.cpp
In file included from /Users/kpai/src/presto/presto-native-execution/_build/debug/presto_cpp/main/thrift/presto_cpp/main/thrift/gen-cpp2/presto_thrift_data.cpp:8:
In file included from /Users/kpai/src/presto/presto-native-execution/_build/debug/presto_cpp/main/thrift/presto_cpp/main/thrift/gen-cpp2/../gen-cpp2/presto_thrift_data.h:11:
In file included from /Users/kpai/src/presto/presto-native-execution/_build/debug/presto_cpp/main/thrift/presto_cpp/main/thrift/gen-cpp2/../gen-cpp2/presto_thrift_types.h:9:
In file included from /usr/local/include/thrift/lib/cpp2/gen/module_types_h.h:28:
In file included from /usr/local/include/folly/container/F14Map.h:37:
In file included from /usr/local/include/folly/container/View.h:20:
/usr/local/include/folly/functional/Invoke.h:22:10: fatal error: 'boost/preprocessor/control/expr_iif.hpp' file not found
#include <boost/preprocessor/control/expr_iif.hpp>
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
[626/679] Building CXX object presto_cpp/main/thrift/CMakeFiles/presto_thrift-cpp2-obj.dir/presto_cpp/main/thrift/gen-cpp2/presto_thrift_metadata.cpp.o
FAILED: presto_cpp/main/thrift/CMakeFiles/presto_thrift-cpp2-obj.dir/presto_cpp/main/thrift/gen-cpp2/presto_thrift_metadata.cpp.o 
ccache /Library/Developer/CommandLineTools/usr/bin/c++ -DFOLLY_HAVE_INT128_T=1 -I/Users/kpai/src/presto/presto-native-execution/. -I/Users/kpai/src/presto/presto-native-execution/_build/debug -I/Users/kpai/src/presto/presto-native-execution/velox -I/Users/kpai/src/presto/presto-native-execution/_build/debug/velox -I/Users/kpai/src/presto/presto-native-execution/_build/debug/_deps/gtest-src/googletest/include -isystem /usr/local/include -isystem /Users/kpai/homebrew/opt/openssl@1.1/include -isystem /usr/local -isystem /usr/local/include/antlr4-runtime -mcpu=apple-m1+crc -std=c++17 -fvisibility=hidden -fvisibility=hidden -fvisibility-inlines-hidden -Werror -Wno-nullability-completeness -Wno-deprecated-declarations -Wreorder -g -std=gnu++17 -arch arm64 -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX13.3.sdk -MD -MT presto_cpp/main/thrift/CMakeFiles/presto_thrift-cpp2-obj.dir/presto_cpp/main/thrift/gen-cpp2/presto_thrift_metadata.cpp.o -MF presto_cpp/main/thrift/CMakeFiles/presto_thrift-cpp2-obj.dir/presto_cpp/main/thrift/gen-cpp2/presto_thrift_metadata.cpp.o.d -o presto_cpp/main/thrift/CMakeFiles/presto_thrift-cpp2-obj.dir/presto_cpp/main/thrift/gen-cpp2/presto_thrift_metadata.cpp.o -c /Users/kpai/src/presto/presto-native-execution/_build/debug/presto_cpp/main/thrift/presto_cpp/main/thrift/gen-cpp2/presto_thrift_metadata.cpp
In file included from /Users/kpai/src/presto/presto-native-execution/_build/debug/presto_cpp/main/thrift/presto_cpp/main/thrift/gen-cpp2/presto_thrift_metadata.cpp:7:
In file included from /usr/local/include/thrift/lib/cpp2/gen/module_metadata_cpp.h:23:
In file included from /usr/local/include/thrift/lib/cpp2/gen/module_metadata_h.h:22:
In file included from /usr/local/include/thrift/lib/thrift/gen-cpp2/metadata_types.h:9:
In file included from /usr/local/include/thrift/lib/cpp2/gen/module_types_h.h:28:
In file included from /usr/local/include/folly/container/F14Map.h:37:
In file included from /usr/local/include/folly/container/View.h:20:
/usr/local/include/folly/functional/Invoke.h:22:10: fatal error: 'boost/preprocessor/control/expr_iif.hpp' file not found
#include <boost/preprocessor/control/expr_iif.hpp>
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
[627/679] Building CXX object presto_cpp/main/thrift/CMakeFiles/presto_thrift-cpp2-obj.dir/presto_cpp/main/thrift/gen-cpp2/presto_thrift_types.cpp.o
FAILED: presto_cpp/main/thrift/CMakeFiles/presto_thrift-cpp2-obj.dir/presto_cpp/main/thrift/gen-cpp2/presto_thrift_types.cpp.o 
ccache /Library/Developer/CommandLineTools/usr/bin/c++ -DFOLLY_HAVE_INT128_T=1 -I/Users/kpai/src/presto/presto-native-execution/. -I/Users/kpai/src/presto/presto-native-execution/_build/debug -I/Users/kpai/src/presto/presto-native-execution/velox -I/Users/kpai/src/presto/presto-native-execution/_build/debug/velox -I/Users/kpai/src/presto/presto-native-execution/_build/debug/_deps/gtest-src/googletest/include -isystem /usr/local/include -isystem /Users/kpai/homebrew/opt/openssl@1.1/include -isystem /usr/local -isystem /usr/local/include/antlr4-runtime -mcpu=apple-m1+crc -std=c++17 -fvisibility=hidden -fvisibility=hidden -fvisibility-inlines-hidden -Werror -Wno-nullability-completeness -Wno-deprecated-declarations -Wreorder -g -std=gnu++17 -arch arm64 -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX13.3.sdk -MD -MT presto_cpp/main/thrift/CMakeFiles/presto_thrift-cpp2-obj.dir/presto_cpp/main/thrift/gen-cpp2/presto_thrift_types.cpp.o -MF presto_cpp/main/thrift/CMakeFiles/presto_thrift-cpp2-obj.dir/presto_cpp/main/thrift/gen-cpp2/presto_thrift_types.cpp.o.d -o presto_cpp/main/thrift/CMakeFiles/presto_thrift-cpp2-obj.dir/presto_cpp/main/thrift/gen-cpp2/presto_thrift_types.cpp.o -c /Users/kpai/src/presto/presto-native-execution/_build/debug/presto_cpp/main/thrift/presto_cpp/main/thrift/gen-cpp2/presto_thrift_types.cpp
In file included from /Users/kpai/src/presto/presto-native-execution/_build/debug/presto_cpp/main/thrift/presto_cpp/main/thrift/gen-cpp2/presto_thrift_types.cpp:7:
In file included from /Users/kpai/src/presto/presto-native-execution/_build/debug/presto_cpp/main/thrift/presto_cpp/main/thrift/gen-cpp2/../gen-cpp2/presto_thrift_types.h:9:
In file included from /usr/local/include/thrift/lib/cpp2/gen/module_types_h.h:28:
In file included from /usr/local/include/folly/container/F14Map.h:37:
In file included from /usr/local/include/folly/container/View.h:20:
/usr/local/include/folly/functional/Invoke.h:22:10: fatal error: 'boost/preprocessor/control/expr_iif.hpp' file not found
#include <boost/preprocessor/control/expr_iif.hpp>
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    ```

    4) Sodium_SOURCE=BUNDLED , also unfortunately doesnt work. 
    
    ```
    -- Setting Sodium source to BUNDLED
CMake Error at velox/CMake/ResolveDependency.cmake:43 (include):
  include could not find requested file:

    sodium
Call Stack (most recent call first):
  velox/CMake/ResolveDependency.cmake:81 (build_dependency)
  CMakeLists.txt:162 (resolve_dependency)
```

5) See the same error as above for fizz, wangle, FBThrift
6) re2 gives me the following error , which looks like a cmake script error: 
```
-- Building re2 from source
CMake Error at velox/CMake/resolve_dependency_modules/re2.cmake:38 (if):
  if given arguments:

    "STREQUAL" "BUNDLED"

  Unknown arguments specified
Call Stack (most recent call first):
  velox/CMake/ResolveDependency.cmake:43 (include)
  velox/CMake/ResolveDependency.cmake:81 (build_dependency)
  velox/CMakeLists.txt:411 (resolve_dependency)
```

nkaretnikov commented:
@kgpai Thanks for giving it a try! Let me summarize:
- this PR adds the `resolve_dependency` macro: you can select SYSTEM or BUNDLED
- SYSTEM relies on `find_package`, which needs `Find<package>.cmake`. These files are usually provided by your system, but I wrote three additional files to make this mode work for antlr, re2, proxygen.
- BUNDLED relies on `<package>.cmake` files, list of available packages [here](https://github.com/prestodb/presto/pull/20732#discussion_r1317350516). I wrote no additional files for this mode. _When we discussed this last time, it was considered out of scope._
- If there is no `<package>.cmake` file, BUNDLED won't find the package
- If there are dependency conflicts or syntax issues in the `<package>.cmake` file, BUNDLED will also fail.

I've tried building some packages via BUNDLED:
- fmt worked right away
- Boost and re2 worked with a bit of effort
- gflags and glog didn't work. For these two, there's additional logic in velox's CMakeLists.txt. Might be tricky to build from source, I haven't investigated.
- proxygen didn't work. This is _expected_ because there's no `<package>.cmake` for it.

I'm attaching the diffs below with the Dockerfile I use for testing.

# Dockerfile
```dockerfile
FROM ubuntu:22.04

ARG DEBIAN_FRONTEND=noninteractive

RUN apt-get update
RUN apt-get install -y git build-essential cmake sudo python3-pip unzip

# Uses my fork for now
RUN git clone -b resolve-dependency-cmake https://github.com/nkaretnikov/presto && \
  cd presto/presto-native-execution && \
  make submodules && \
  bash ./scripts/setup-ubuntu.sh
```

Build it:
```bash
docker build --tag presto:latest - < Dockerfile && docker run -it presto:latest /bin/bash
```

In docker (you'll need to re-run this step after every change to `CMakeLists.txt` - I use `vim` in docker to edit it, because it's cloned from GitHub and not mapped from local filesystem):
```bash
cd /presto/presto-native-execution/
mkdir _build && \
  cd _build && \
  cmake .. && \
  cmake --build . --config Release -j$(($(nproc) * 3/4)) && \
  ctest || true
```

# Bundled Boost

Needs Threads.

```diff
diff --git a/presto-native-execution/CMakeLists.txt b/presto-native-execution/CMakeLists.txt
index d1fcc05150..8f34aba0c1 100644
--- a/presto-native-execution/CMakeLists.txt
+++ b/presto-native-execution/CMakeLists.txt
@@ -132,7 +132,9 @@ set(BOOST_INCLUDE_LIBRARIES
     system
     thread)
 set(Boost_USE_MULTITHREADED ON)
-set_source(Boost)
+find_package(Threads)
+set(Boost_SOURCE BUNDLED)
+# set_source(Boost)
 resolve_dependency(Boost 1.66.0 COMPONENTS ${BOOST_INCLUDE_LIBRARIES})

 set_source(gflags)
```

In the logs: `Using Boost - Bundled`.

# Bundled fmt

```diff
diff --git a/presto-native-execution/CMakeLists.txt b/presto-native-execution/CMakeLists.txt
index d1fcc05150..891a71f30d 100644
--- a/presto-native-execution/CMakeLists.txt
+++ b/presto-native-execution/CMakeLists.txt
@@ -144,7 +144,8 @@ resolve_dependency(antlr4-runtime)
 set_source(glog)
 resolve_dependency(glog)

-set_source(fmt)
+# set_source(fmt)
+set(fmt_SOURCE BUNDLED)
 resolve_dependency(fmt)

 add_compile_definitions(FOLLY_HAVE_INT128_T=1)
```

In the logs: `Building fmt from source.`

# Bundled re2

Needs ICU_SOURCE defined, but it's not a hard dependency.

The warning comes from `velox/CMake/resolve_dependency_modules/re2.cmake:38`:

```cmake
if(${ICU_SOURCE} STREQUAL "BUNDLED")
```
where the code expects the variable to be defined. If it's not, you get a syntax error.

```diff
diff --git a/presto-native-execution/CMakeLists.txt b/presto-native-execution/CMakeLists.txt
index d1fcc05150..d6ea6c1775 100644
--- a/presto-native-execution/CMakeLists.txt
+++ b/presto-native-execution/CMakeLists.txt
@@ -166,7 +166,9 @@ set_source(wangle)
 resolve_dependency(wangle)
 set_source(proxygen)
 resolve_dependency(proxygen)
-set_source(re2)
+# set_source(re2)
+set(ICU_SOURCE notfound)
+set(re2_SOURCE BUNDLED)
 resolve_dependency(re2)
 set_source(FBThrift)
 resolve_dependency(FBThrift)
```

In the logs: `Building re2 from source`.

# Bundled proxygen

Doesn't work because there's no `<package>.cmake` file for it (expected).

nkaretnikov commented:
The SYSTEM install method works. However, to make the BUNDLED mode work, this required packaging additional dependencies that are only present in presto, which would be time consuming. This was abandoned in favor of installing via conda https://github.com/prestodb/presto/pull/20842. Closing.
